File: src/gitops.ts
import { getProjectDirFromRemote } from "./project";
import to from "await-to-js";
import fs from "fs-extra";
import chalk from "chalk";
import { Config, Project } from "./types/config";
import { ErrorWithHint, ToChildProcessOutput } from "./types/utils";
import { applyPromiseToEntriesWithProgressBar } from "./progress";
import * as utils from "./utils";
import { AssertionError } from "assert";
import tildify from "tildify";

type LogFn = (arg: string | ErrorWithHint) => void;

export async function hasLocalChanges(dir: string): Promise<boolean> {
	const [err, res]: ToChildProcessOutput = await to(
		utils.spawn("git", ["status", "--porcelain"], { cwd: dir, encoding: "utf8" }),
	);
	if (err || !res) {
		throw new ErrorWithHint(chalk`{yellow ${tildify(dir)}} {red failed} to check for local changes ${err?.stderr}`);
	}
	return `${res.stdout}`.trim().length !== 0;
}

async function pull(dir: string, currentBranch: string, log: LogFn) {
	const pcpPromise = utils.spawn("git", ["pull", "--ff-only"], { cwd: dir, encoding: "utf8" });
	const [err, res]: ToChildProcessOutput = await to(pcpPromise);

	if (err || !res) {
		if (`${err?.stderr}`.trim().includes("There is no tracking information for the current branch")) {
			log(chalk`{cyan ${currentBranch}} {red doesn't have a remote origin} in {cyan ${tildify(dir)}}`);
		} else if (`${err?.stderr}`.trim().includes(`Your configuration specifies to merge with the ref`)) {
			log(chalk`{cyan ${currentBranch}} {red no such ref could be fetched} in {cyan ${tildify(dir)}}`);
		} else {
			log(
				chalk`{cyan ${currentBranch}} {red conflicts} with {magenta origin/${currentBranch}} in {cyan ${tildify(dir)}}`,
			);
		}
		return false;
	}

	const msg = `${res.stdout}`.trim();
	if (msg === "Already up to date.") {
		log(chalk`{cyan ${currentBranch}} is up to date with {magenta origin/${currentBranch}} in {cyan ${tildify(dir)}}`);
	} else {
		log(chalk`{cyan ${currentBranch}} pulled changes from {magenta origin/${currentBranch}} in {cyan ${tildify(dir)}}`);
	}
	return true;
}

async function mergeError(dir: string, currentB: string, defaultB: string, log: LogFn) {
	log(chalk`{yellow ${currentB}} merge with {magenta origin/${defaultB}} {red failed} in {cyan ${tildify(dir)}}`);

	const pcpPromise = utils.spawn("git", ["merge", "--abort"], { cwd: dir, encoding: "utf8" });
	const [err] = await to(pcpPromise);
	if (err) {
		log(chalk`{yellow ${currentB}} merge --abort also {red failed} in {cyan ${tildify(dir)}}`);
	}
}

async function merge(dir: string, currentB: string, defaultBranch: string, log: LogFn) {
	const pcpPromise = utils.spawn("git", ["merge", `origin/${defaultBranch}`], { cwd: dir, encoding: "utf8" });
	const [err, res] = await to(pcpPromise);

	if (err || !res) {
		return mergeError(dir, currentB, defaultBranch, log);
	}

	const msg = `${res.stdout}`.trim();
	if (msg === "Already up to date.") {
		log(chalk`{cyan ${currentB}} is up to date with {magenta origin/${defaultBranch}} in {cyan ${tildify(dir)}}`);
	} else {
		log(chalk`{yellow {cyan ${currentB}} was merged with {magenta origin/${defaultBranch}} in {cyan ${tildify(dir)}}}`);
	}
}

async function clone(cwd: string, remote: string, dir: string, log: LogFn) {
	const pcpPromise = utils.spawn("git", ["clone", remote, dir], { cwd, encoding: "utf8" });
	const [err, res]: ToChildProcessOutput = await to(pcpPromise);
	if (err || !res) {
		if (err?.stderr?.includes("Permission denied")) {
			const errorMessage = `Permission denied to clone ${remote}`;
			log(new ErrorWithHint(errorMessage));
			return;
		}
		log(new ErrorWithHint(chalk`{yellow ${remote}} {red failed} in {cyan ${tildify(dir)}} \n${err?.stderr}`));
		return;
	}
	log(chalk`{gray ${remote}} was cloned to {cyan ${tildify(dir)}}`);
}

export async function logHowFarBehindDefaultBranch(
	cwd: string,
	currentBranch: string,
	defaultBranch: string,
	log: LogFn,
) {
	const [err, res]: ToChildProcessOutput = await to(
		utils.spawn("git", ["rev-list", "--count", "--left-right", `${currentBranch}..origin/${defaultBranch}`], {
			cwd,
			encoding: "utf8",
		}),
	);
	if (err || !res) {
		return;
	}
	const splitted = res.stdout.toString().trim().split("\t");
	if (splitted.length !== 2) {
		return;
	}

	if (splitted[1] !== "0") {
		log(
			chalk`{yellow ${currentBranch}} is {red ${splitted[1]}} commits behind {cyan ${defaultBranch}} in {cyan ${tildify(
				cwd,
			)}}`,
		);
	}
	if (splitted[0] !== "0") {
		log(
			chalk`{yellow ${currentBranch}} is {green ${
				splitted[0]
			}} commits ahead of {cyan ${defaultBranch}} in {cyan ${tildify(cwd)}}`,
		);
	}
}

export async function gitops(
	cwd: string,
	projectObj: Project,
	autoMerge: boolean,
): Promise<(string | ErrorWithHint)[]> {
	const logs: (string | ErrorWithHint)[] = [];
	const log = (arg: string | ErrorWithHint) => {
		logs.push(arg);
	};
	const remote = projectObj["remote"];
	const defaultBranch = projectObj["default_branch"];
	const dir = getProjectDirFromRemote(cwd, remote);

	if (!(await fs.pathExists(`${dir}`))) {
		await clone(cwd, remote, dir, log);
		return logs;
	}

	const pcpPromise = utils.spawn("git", ["branch", "--show-current"], { cwd: dir, encoding: "utf8" });
	const [err, res]: ToChildProcessOutput = await to(pcpPromise);

	if (err || !res) {
		log(new ErrorWithHint(chalk`{yellow ${remote}} {red failed} in {cyan ${tildify(dir)}} ${err}`));
		return logs;
	}

	const currentBranch = `${res.stdout}`.trim();

	let localChanges: boolean;
	try {
		localChanges = await hasLocalChanges(dir);
	} catch (error) {
		if (error instanceof ErrorWithHint) {
			log(error);
			return logs;
		}
		throw error;
	}

	if (localChanges) {
		log(chalk`{yellow ${currentBranch}} has local changes in {cyan ${dir}}`);
	} else if (currentBranch === defaultBranch) {
		await pull(dir, currentBranch, log);
	} else {
		if (await pull(dir, currentBranch, log)) {
			if (autoMerge) {
				await merge(dir, currentBranch, defaultBranch, log);
			}
		}
	}
	await logHowFarBehindDefaultBranch(dir, currentBranch, defaultBranch, log);
	return logs;
}

export function handleGitopsResults(projectNames: string[], logs: (string | ErrorWithHint)[][]) {
	let errorCount = 0;
	for (const [i, projectName] of projectNames.entries()) {
		const isError = logs[i].filter((log) => log instanceof ErrorWithHint).length > 0;

		if (!isError) {
			console.log(chalk`┌─ {green {bold ${projectName}}}`);
		} else {
			errorCount++;
			console.log(chalk`┌─ {red {bold ${projectName}}}`);
		}

		for (const [j, log] of logs[i].entries()) {
			let formattedLog = "";
			if (log instanceof ErrorWithHint) {
				formattedLog = chalk`{red ${log.hint}}`;
			} else {
				formattedLog = log;
			}

			console.log(`${j === logs[i].length - 1 ? "└" : "├"}─── ${formattedLog}`);
		}
	}

	if (errorCount > 0) {
		throw new AssertionError({ message: "At least one git operation failed" });
	}
}

export async function fromConfig(cnf: Config, autoMerge: boolean) {
	utils.printHeader("Git Operations");
	const fn = (arg: Project) => gitops(cnf.cwd, arg, autoMerge);
	const result = await applyPromiseToEntriesWithProgressBar("git-operations", Object.entries(cnf.projects), fn);
	console.log();
	handleGitopsResults(Object.keys(cnf.projects), result);
}

File: internal/gitops.go
package internal

import (
	"context"
	"fmt"
	"gitte/config"
	"gitte/executor"
	"os"
	"path/filepath"
	"regexp"
)

func GitOps(ctx context.Context, cwd string, gitteConfig config.GitteConfig) error {
	tasks := []executor.Task{}
	for name, pc := range gitteConfig.Projects {
		tasks = append(tasks, executor.Task{
			Name: fmt.Sprintf("gitops-%s", name),
			ExecuteFn: func(ctx context.Context, name string, oh executor.OutputHandler) error {
				return gitopsProject(ctx, cwd, pc)
			},
		})
	}

	return executor.NewExecutor(tasks).Execute(ctx)
}

func gitopsProject(ctx context.Context, cwd string, pc config.ProjectConfig) error {
	relativeDirectory, err := getProjectDirFromRemote(pc)
	if err != nil {
		return err
	}

	projectPath := filepath.Join(cwd, relativeDirectory)

	// Check if directory exists
	if _, err := os.Stat(projectPath); os.IsNotExist(err) {
		// Directory does not exist, clone the repository
		fmt.Println("Cloning repository:", pc.Remote, "into", projectPath)
		return clone(ctx, cwd, pc)
	}

	localChanges, err := hasLocalChanges(projectPath)
	if err != nil {
		return err
	}

	if localChanges {
		return fmt.Errorf("local changes detected in project at '%s'. Please commit or stash them before pulling updates", projectPath)
	}

	err = pull(ctx, projectPath)
	if err != nil {
		return err
	}

	return nil // TODO log how far behind current branch is from default branch
}

func clone(ctx context.Context, cwd string, pc config.ProjectConfig) error {
	relativeDirectory, err := getProjectDirFromRemote(pc)
	if err != nil {
		return err
	}

	res, err := executor.ExecuteSyncInDir(cwd, "git", "clone", pc.Remote, relativeDirectory)
	if err != nil {
		return fmt.Errorf("failed to execute git clone command: %w", err)
	}

	if res.ExitCode != 0 {
		// If res.Stderr contains "Permission denied" return a specific error
		if regexp.MustCompile(`(?i)permission denied`).Match(res.Stderr) {
			return fmt.Errorf("permission denied when trying to clone repository '%s'. Please check your SSH keys and access rights", pc.Remote)
		}
		return fmt.Errorf("git clone command failed with exit code %d: \n%s", res.ExitCode, string(res.Stderr))
	}

	return nil
}

func getCurrentBranch(cwd string) (string, error) {
	res, err := executor.ExecuteSyncInDir(cwd, "git", "branch", "--show-current")
	if err != nil {
		return "", fmt.Errorf("failed to execute git branch command: %w", err)
	}

	if res.ExitCode != 0 {
		return "", fmt.Errorf("git branch command failed with exit code %d: \n%s", res.ExitCode, string(res.Stderr))
	}

	return string(res.Stdout), nil
}

func hasLocalChanges(cwd string) (bool, error) {
	res, err := executor.ExecuteSyncInDir(cwd, "git", "status", "--porcelain")
	if err != nil {
		return false, fmt.Errorf("failed to execute git status command: %w", err)
	}

	if res.ExitCode != 0 {
		return false, fmt.Errorf("git status command failed with exit code %d: \n%s", res.ExitCode, string(res.Stderr))
	}
	return len(res.Stdout) > 0, nil
}

func pull(ctx context.Context, cwd string) error {
	res, err := executor.ExecuteSyncInDir(cwd, "git", "pull", "--ff-only")
	if err != nil {
		return fmt.Errorf("failed to execute git pull command: %w", err)
	}

	if res.ExitCode != 0 {
		if regexp.MustCompile(`(?i)there is no tracking information for the current branch`).Match(res.Stderr) {
			return fmt.Errorf("the current branch doesn't have a remote origin in '%s'", cwd)
		} else if regexp.MustCompile(`(?i)your configuration specifies to merge with the ref`).Match(res.Stderr) {
			return fmt.Errorf("no such ref could be fetched in '%s'", cwd)
		} else if regexp.MustCompile(`(?i)conflicts with`).Match(res.Stderr) {
			return fmt.Errorf("the current branch has conflicts with the remote branch in '%s'", cwd)
		} else if regexp.MustCompile(`(?i)can't be fast-forwarded`).Match(res.Stderr) {
			return fmt.Errorf("the current branch cannot be fast-forwarded in '%s'", cwd)
		}

		return fmt.Errorf("git pull command failed with exit code %d: \n%s", res.ExitCode, string(res.Stderr))
	}

	return nil
}

func getProjectDirFromRemote(pc config.ProjectConfig) (string, error) {
	regex := regexp.MustCompile(`git@.*?:(.*)?\.git`)
	match := regex.FindStringSubmatch(pc.Remote)

	firstMatch := ""
	if len(match) > 1 {
		firstMatch = match[1]
	}

	if firstMatch == "" {
		return "", fmt.Errorf("'%' is not a valid git ssh url. Use git@gitlab.com:example/cego.git syntax", pc.Remote)
	}

	return firstMatch, nil
}

func getCommitCountBehindDefaultBranch(cwd string, branch string, defaultBranch string) (int, error) {
	res, err := executor.ExecuteSyncInDir(cwd, "git", "rev-list", "--count", "--left-right", fmt.Sprintf("%s...origin/%s", branch, defaultBranch))
	if err != nil {
		return 0, fmt.Errorf("failed to execute git rev-list command: %w", err)
	}

	if res.ExitCode != 0 {
		return 0, fmt.Errorf("git rev-list command failed with exit code %d: \n%s", res.ExitCode, string(res.Stderr))
	}

	var behindCount int
	_, err = fmt.Sscanf(string(res.Stdout), "%*d\t%d", &behindCount)
	if err != nil {
		return 0, fmt.Errorf("failed to parse git rev-list output: %w", err)
	}

	return behindCount, nil
}
